src/server/pfs/server/api_server.go

```diff
@@ -390,42 +411,74 @@ func (a *apiServer) PutFile(putFileServer pfsclient.API_PutFileServer) (retErr e
 			if _, err := putFileClient.CloseAndRecv(); err != nil {
 				return err
 			}
+			return nil
+		}
+		defer func() {
+			if _, err := putFileClient.CloseAndRecv(); err != nil && retErr == nil {
+				retErr = err
+			}
+		}()
+		if err := putFileClient.Send(request); err != nil {
+			return err
+		}
+		for {
+			request, err := putFileServer.Recv()
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				return err
+			}
+			if err := putFileClient.Send(request); err != nil {
+				return err
+			}
 		}
 		return nil
 	}
-	clientConn, err := a.getClientConnForFile(request.File, a.version)
-	if err != nil {
-		return err
-	}
-	putFileClient, err := pfsclient.NewInternalAPIClient(clientConn).PutFile(ctx)
-	if err != nil {
-		return err
+
+	for _, req := range requests {
+		wg.Add(1)
+		go func(req *pfsclient.PutFileRequest) {
+			defer wg.Done()
+			err := sendReq(req)
+			if err != nil {
+				select {
+				case errCh <- err:
+					// error reported
+				default:
+					// not the first error
+				}
+				return
+			}
+		}(req)
 	}
-	defer func() {
-		if _, err := putFileClient.CloseAndRecv(); err != nil && retErr == nil {
-			retErr = err
-		}
-	}()
-	if err := putFileClient.Send(request); err != nil {
+
+	wg.Wait()
+
+	select {
+	case err := <-errCh:
 		return err
+	default:
+		return nil
 	}
+}
```

Comment:
The abstraction in this function is a little bit weird, you're first creating a slice of requests, then a closure to handle each of those requests, then iterating through the requests and calling the closure on them. I think it would be a lot easier to just have the first loop dispatch the requests inline. Then handle the original request afterward, no closures or intermediate slices needed.